src/solvers/base_solver.cpp

```diff
@@ -35,143 +35,76 @@ BaseSolver::BaseSolver(MPI_Comm comm, int n, int p) : BaseSolver(comm)
+  // TODO: Raise an issue against MFEM
+  // Leave this explicit non-const action in as a stopgap
+  const_cast<mfem::ParFiniteElementSpace&>(fes).GetEssentialTrueDofs(bc.markers, bc.true_dofs, component);
 
-  ess_bdr_.push_back(bc);
+  bc.coef = ess_bdr_coef;
+  ess_bdr_.push_back(std::move(bc));
 }
 
-void BaseSolver::setTrueDofs(const mfem::Array<int>& true_dofs, std::shared_ptr<mfem::Coefficient> ess_bdr_coef)
+void BaseSolver::setTrueDofs(const mfem::Array<int>& true_dofs, serac::BoundaryCondition::Coef ess_bdr_coef,
+                             int component)
 {
-  auto bc = std::make_shared<serac::BoundaryCondition>();
+  serac::BoundaryCondition bc;
 
-  bc->markers.SetSize(0);
+  bc.markers.SetSize(0);
 
-  bc->true_dofs = true_dofs;
+  bc.true_dofs = true_dofs;
 
-  bc->scalar_coef = ess_bdr_coef;
+  bc.coef = ess_bdr_coef;
 
-  ess_bdr_.push_back(bc);
+  bc.component = component;
+  ess_bdr_.push_back(std::move(bc));
 }
 
-void BaseSolver::setNaturalBCs(const std::set<int>& nat_bdr, std::shared_ptr<mfem::Coefficient> nat_bdr_coef,
-                               int component)
+void BaseSolver::setNaturalBCs(const std::set<int>& nat_bdr, serac::BoundaryCondition::Coef nat_bdr_coef, int component)
 {
-  auto bc = std::make_shared<serac::BoundaryCondition>();
+  serac::BoundaryCondition bc;
 
-  bc->markers.SetSize(state_.front()->mesh->bdr_attributes.Max());
-  bc->markers = 0;
+  bc.markers.SetSize(state_.front()->mesh->bdr_attributes.Max());
+  bc.markers = 0;
 
   for (int attr : nat_bdr) {
     SLIC_ASSERT_MSG(attr <= bc->markers.Size(), "Attribute specified larger than what is found in the mesh.");
-    bc->markers[attr - 1] = 1;
+    bc.markers[attr - 1] = 1;
   }
-
-  bc->scalar_coef = nat_bdr_coef;
-  bc->component   = component;
-
-  nat_bdr_.push_back(bc);
+  bc.component = component;
+  bc.coef      = nat_bdr_coef;
+  nat_bdr_.push_back(std::move(bc));
 }
 
-void BaseSolver::setState(const std::vector<std::shared_ptr<mfem::Coefficient> >& state_coef)
+void BaseSolver::setState(const std::vector<serac::BoundaryCondition::Coef>& state_coef)
 {
   SLIC_ASSERT_MSG(state_coef.size() == state_.size(), "State and coefficient bundles not the same size.");
 
   for (unsigned int i = 0; i < state_coef.size(); ++i) {
-    state_[i]->gf->ProjectCoefficient(*state_coef[i]);
-  }
-}
-
-void BaseSolver::setState(const std::vector<std::shared_ptr<mfem::VectorCoefficient> >& state_vec_coef)
-{
-  SLIC_ASSERT_MSG(state_vec_coef.size() == state_.size(), "State and coefficient bundles not the same size.");
-
-  for (unsigned int i = 0; i < state_vec_coef.size(); ++i) {
-    state_[i]->gf->ProjectCoefficient(*state_vec_coef[i]);
+    // The generic lambda parameter, auto&&, allows the component type (mfem::Coef or mfem::VecCoef)
+    // to be deduced, and the appropriate version of ProjectCoefficient is dispatched.
+    std::visit([this, i](auto&& coef) { state_[i]->gf->ProjectCoefficient(*coef); }, state_coef[i]);
```

Comment:
I think now that I've had some time to let `std::visit` percolate in my brain, this is more readable.